### DISCUSSION TITLE
Power Delivery Message Decoding

### DISCUSSION THREAD

**Discussion by topherbuckley** [hathach/tinyusb#2007] [Q&A] [answered]
Does TinyUSB have any Power Delivery support? I'm hoping to find somewhere that can take a PD message as a byte array and give me back something more human readable like you get with protocol analyzers or the like. Hoping to make use of this with an rp2040 if it makes any difference.

**Comment by hathach** [comment]
rp2040 does not support PD physically and tinyusb also does not support PD yet.

**Reply by topherbuckley** [reply]
Alrighty, thanks for the info! Yeah, I have a max77958 chip to handle the PD itself, but a rp2040 is responding to interrupts from that chip. One of which is signaling a PD message has been received/sent. Is there any dev code or wip for PD here yet? Can't promise much time, but happy to add something to it at least as far as decoding PD messages (since I need to implement that anyway).

**Reply by hathach** [reply]
unfortunately, there is no effort so far with PD since I don't have any hw that could actual do PD.

**Reply by topherbuckley** [reply]
Alrighty. Thanks for the info though. I may have found another repo working on this, but if that flops, I'll see if I can produce anything useful here.

**Comment by manuelbl** [comment, marked as answer]
[USB PD for Arduino](https://github.com/manuelbl/usb-pd-arduino) is an open-source project implementing USB PD communication and a protocol analyzer, without the need for a USB PD controller IC. While it is mainly targeted at STM32 chips and is based on the Arduino framework, the code might still be useful for your project.

Since you already have the code to interact with the lower levels (the MAX chip in your case), you will be mainly interested in the general USB PD classes: `PDMessage` and `USBPDProtocolAnalyzer`. If you also need to implement the USB PD interaction, the classes `PDController` and `PDSink` will be helpful.

*Note: I'm the project author.*
